### PR TITLE
Migrate desugaring expression to iterative optimizer

### PR DESCRIPTION
**Pull request by szymonm**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 1]
```diff
@@ -173,7 +175,13 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                         new PushProjectionThroughExchange()));
 
         builder.add(
-                new DesugaringOptimizer(metadata, sqlParser), // Clean up all the sugar in expressions, e.g. AtTimeZone, must be run before all the other optimizers
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: remain the comment, so it is known that this group of rules is special

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    private Set<Rule<?>> rules = ImmutableSet.of(
+            new ProjectExpressionRewrite(this::rewrite),
+            new AggregationExpressionRewrite(this::rewrite),
+            new FilterExpressionRewrite(this::rewrite),
+            new TableScanExpressionRewrite(this::rewrite),
+            new JoinExpressionRewrite(this::rewrite),
+            new ValuesExpressionRewrite(this::rewrite));
+
+    public static Expression rewrite(Expression expression, Session session, Metadata metadata, SqlParser sqlParser, SymbolAllocator symbolAllocator)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java]: I do not like that you mix expression rewriter with rule set. To me it should be divided into two entities: `DesugarAtTimeZoneRewriter` (expression rewriter) and `DesugarAtTiimeZoneRuleSet`. 

I do not like that fact `PlanOptimizer` is using methods from `Rule`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarLambdaExpression.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        return ImmutableSet.of(
+                new ProjectExpressionRewrite(DesugarLambdaExpression::rewrite),
+                new AggregationExpressionRewrite(DesugarLambdaExpression::rewrite),
+                new FilterExpressionRewrite(DesugarLambdaExpression::rewrite),
+                new TableScanExpressionRewrite(DesugarLambdaExpression::rewrite),
+                new JoinExpressionRewrite(DesugarLambdaExpression::rewrite),
+                new ValuesExpressionRewrite(DesugarLambdaExpression::rewrite));
+    }
+
+    static Expression rewrite(Expression expression, Rule.Context context)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarLambdaExpression.java]: why not private?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLambdaCaptureDesugaringRewriter.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import org.testng.annotations.Test;
+
+import java.util.Map;
+
+import static com.facebook.presto.sql.planner.iterative.rule.LambdaCaptureDesugaringRewriter.rewrite;
+import static com.facebook.presto.sql.planner.iterative.rule.test.PlanBuilder.expression;
+import static java.util.stream.Collectors.toList;
+import static org.testng.Assert.assertEquals;
+
+public class TestLambdaCaptureDesugaringRewriter
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestLambdaCaptureDesugaringRewriter.java]: What about testing for `DesugarAtTimeZone`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java [hunk 5]
```diff
@@ -12,13 +12,16 @@
  * limitations under the License.
  */
 
-package com.facebook.presto.sql.planner;
+package com.facebook.presto.sql.planner.iterative.rule;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: keep such changes as separate commits. Please do not mix logical changes with refactoring changes in the same commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(sqlParser, "sqlParser is null");
+        checkState(!(expression instanceof GroupingOperation), "GroupingOperation should have been re-written to a FunctionCall before execution");
+
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        Map<NodeRef<Expression>, Type> expressionTypes = getExpressionTypes(session, metadata, sqlParser, symbolAllocator.getTypes(), expression, emptyList()); // parameters already replaced
+
+        expression = ExpressionTreeRewriter.rewriteWith(new DesugaringRewriter(expressionTypes), expression);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarAtTimeZone.java]: This class looks like it needs to be renamed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java (hunk 5 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: Sorry for not asking before, but can you please tell me why this package is a better place for this class?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java [hunk 7]
```diff
@@ -33,25 +34,24 @@
...
-        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression, new Context());
+        if (expression instanceof SymbolReference) {
+            return expression;
+        }
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(symbolTypes, symbolAllocator), expression, new Context());
     }
 
+    private LambdaCaptureDesugaringRewriter() {}
+
     private static Expression replaceSymbols(Expression expression, ImmutableMap<Symbol, Symbol> symbolMapping)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: nit: you can use `com.facebook.presto.sql.planner.ExpressionSymbolInliner` instead of this method.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java [hunk 8]
```diff
@@ -202,8 +202,9 @@ private Expression desugar(Expression expression)
             }
             Map<NodeRef<Expression>, Type> expressionTypes = getExpressionTypes(session, metadata, sqlParser, types, expression, emptyList() /* parameters already replaced */);
 
-            expression = new LambdaCaptureDesugaringRewriter(types, symbolAllocator).rewrite(expression);
+            expression = LambdaCaptureDesugaringRewriter.rewrite(expression, symbolAllocator.getTypes(), symbolAllocator);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java]: Maybe passing symbol allocator would be enought?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestLambdaCaptureDesugaringRewriter.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            throws Exception
+    {
+        final Map<Symbol, Type> symbols = ImmutableMap.of(new Symbol("a"), BigintType.BIGINT);
+        final SymbolAllocator allocator = new SymbolAllocator(symbols);
+
+        assertEquals(rewrite(expression("x -> a + x"), symbols, allocator),
+                     new BindExpression(
+                             ImmutableList.of(expression("a")),
+                             new LambdaExpression(
+                                     ImmutableList.of("a_0", "x").stream().map(Identifier::new).map(LambdaArgumentDeclaration::new).collect(toList()),
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestLambdaCaptureDesugaringRewriter.java]: I would write this as:

```
Stream.of("a_0", "x")
    .map(Identifier::new)
    .map(LambdaArgumentDeclaration::new)
    .collect(toImmutableList())
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DesugarAtTimeZoneRewriter.java [hunk 10]
```diff
@@ -42,6 +50,20 @@ public static Expression rewrite(Expression expression, Map<NodeRef<Expression>,
 
     private DesugarAtTimeZoneRewriter() {}
 
+    public static Expression rewrite(Expression expression, Session session, Metadata metadata, SqlParser sqlParser, SymbolAllocator symbolAllocator)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(sqlParser, "sqlParser is null");
+        checkState(!(expression instanceof GroupingOperation), "GroupingOperation should have been re-written to a FunctionCall before execution");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DesugarAtTimeZoneRewriter.java]: now it became a `checkArgument`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java [hunk 11]
```diff
@@ -194,15 +189,8 @@ public PlanNode visitApply(ApplyNode node, RewriteContext<Void> context)
 
         private Expression desugar(Expression expression)
         {
-            checkState(!(expression instanceof GroupingOperation), "GroupingOperation should have been re-written to a FunctionCall before execution");
-
-            if (expression instanceof SymbolReference) {
-                return expression;
-            }
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DesugaringOptimizer.java]: I think that above `if` and `checkState` should remain here. `if` and `checkState` apply to both rewrites,.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java (hunk 5 above)
Comment (szymonm) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: This rewriter is used by deprecated `DesugarignOptimizer` in `planner.optimizers` and `DesugarLambdaExpression` in `planner.iterative.rule`. After completely moving to the iterative optimizer, I don't think it's gonna be used outside of rules, so I moved it there.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java [hunk 12]
```diff
@@ -33,25 +34,24 @@
...
-    private final Map<Symbol, Type> symbolTypes;
-    private final SymbolAllocator symbolAllocator;
-
-    public LambdaCaptureDesugaringRewriter(Map<Symbol, Type> symbolTypes, SymbolAllocator symbolAllocator)
+    public static Expression rewrite(Expression expression, Map<Symbol, Type> symbolTypes, SymbolAllocator symbolAllocator)
     {
-        this.symbolTypes = requireNonNull(symbolTypes, "symbolTypes is null");
-        this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
-    }
+        checkState(!(expression instanceof GroupingOperation), "GroupingOperation should have been re-written to a FunctionCall before execution");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: It was not here before refactor. Moreover grouping could hidden somewhere in the tree, like for  `GROUPING(...) + 1`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java [hunk 13]
```diff
@@ -33,25 +34,24 @@
...
 
-    public Expression rewrite(Expression expression)
-    {
-        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression, new Context());
+        if (expression instanceof SymbolReference) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: is this needed? not much will be executed for such expression

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 14]
```diff
@@ -220,7 +220,7 @@ public Expression rewriteExpression(Expression node, Void context, ExpressionTre
         analyzer.analyze(rewrite, Scope.create());
 
         // remove syntax sugar
-        rewrite = ExpressionTreeRewriter.rewriteWith(new DesugaringRewriter(analyzer.getExpressionTypes()), rewrite);
+        rewrite = DesugarAtTimeZoneRewriter.rewrite(rewrite, analyzer.getExpressionTypes());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: should be the other desugar rewrite be used here as well?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DesugarAtTimeZoneRewriter.java [hunk 15]
```diff
@@ -42,6 +50,20 @@ public static Expression rewrite(Expression expression, Map<NodeRef<Expression>,
 
     private DesugarAtTimeZoneRewriter() {}
 
+    public static Expression rewrite(Expression expression, Session session, Metadata metadata, SqlParser sqlParser, SymbolAllocator symbolAllocator)
+    {
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(sqlParser, "sqlParser is null");
+        checkArgument(!(expression instanceof GroupingOperation), "GroupingOperation should have been re-written to a FunctionCall before execution");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DesugarAtTimeZoneRewriter.java]: Why grouping causes issues? I thought that it is irrelevant.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 14 above)
Comment (szymonm) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Can't figure it out, we can talk about it outside of the PR. Here, I left the same semantics as before.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java (hunk 13 above)
Comment (szymonm) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: Right, we don't need it here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java (hunk 12 above)
Comment (szymonm) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: It was checked in the DesugaringOptimizer before. Grouping can contain only column references so we can ommit desugaring for it. And also, it should have been rewritten to function call already in QueryPlanner.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java [hunk 16]
```diff
@@ -104,7 +87,10 @@ public Expression rewriteLambdaExpression(LambdaExpression node, Context context
                 newLambdaArguments.add(new LambdaArgumentDeclaration(new Identifier(extraSymbol.getName())));
             }
             newLambdaArguments.addAll(node.getArguments());
-            Expression rewrittenExpression = new LambdaExpression(newLambdaArguments.build(), replaceSymbols(rewrittenBody, captureSymbolToExtraSymbol.build()));
+
+            ImmutableMap<Symbol, Symbol> symbolsMap = captureSymbolToExtraSymbol.build();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/LambdaCaptureDesugaringRewriter.java]: `s/ImmutableMap/Map/`
